packages/data-point/lib/reducer-types/reducer-object/factory.js

```diff
@@ -26,42 +32,61 @@ function isType (source) {
 
 module.exports.isType = isType
 
+/**
+ * @return {Object}
+ */
+function newProps () {
+  return {
+    constants: {},
+    reducers: []
+  }
+}
+
+module.exports.newProps = newProps
+
 /**
  * @param {Function} createReducer
  * @param {Object} source
- * @param {Array} path
- * @param {Array} props
+ * @param {Array} stack
+ * @param {Object} props
  * @returns {Array}
  */
-function getReducerProps (createReducer, source, path = [], props = []) {
+function getProps (createReducer, source, stack = [], props = newProps()) {
   for (let key of Object.keys(source)) {
+    const path = stack.concat(key)
     const value = source[key]
-    const fullPath = path.concat(key)
     if (isPlainObject(value)) {
       // NOTE: recursive call
-      getReducerProps(createReducer, value, fullPath, props)
+      getProps(createReducer, value, path, props)
       continue
     }
 
-    props.push({
-      path: fullPath,
-      reducer: createReducer(value)
-    })
+    const reducer = createReducer(value)
+    if (reducer.type === 'ReducerConstant') {
+      set(props.constants, path, reducer.value)
+    } else {
+      props.reducers.push({ path, reducer })
+    }
   }
 
   return props
 }
 
-module.exports.getReducerProps = getReducerProps
+module.exports.getProps = getProps
 
 /**
  * @param {Function} createReducer
  * @param {Object} source
  * @returns {reducer}
  */
 function create (createReducer, source = {}) {
+  const props = getProps(createReducer, source)
+
   const reducer = new ReducerObject()
-  reducer.props = getReducerProps(createReducer, source)
+  reducer.isEmpty = isEmpty(props.constants) && isEmpty(props.reducers)
+  reducer.source = Function(`return ${JSON.stringify(props.constants)}`) // eslint-disable-line no-new-func
```

Comment:
not confident about doing this - the idea is that it creates a new object whenever `reducer.source` is called (instead of returning the same object every time), and it does that without calling something like `_.cloneDeep` at run time.

one problem is that it means you can't put a function inside of a `constant` when it's used in a reducer object. it's an edge case, but the function here would get turned into a useless string:

```js
const reducer = {
  a: constant(() => 'return something here')
}
```

Reply:
interesting,  this can be a deal breaker for Request’s options.parse and other props that can be a function

prob we should consider using immer?

Reply:
For now I plan to remove this line and open a follow up issue to make constants immutable. However, removing that line caused a circular dependency in the "branch/leaf nesting" test, so I need to look into that.

Reply:
I am now using `_.cloneDeep`, but I can open an issue to see about replacing that with immer or some other approach